### PR TITLE
Measure performance in logs

### PR DESCRIPTION
**Pull request by pgrill**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -309,6 +309,14 @@ All the configurations are grouped in *sections* and some of them can vary depen
...
   - `MaxThreadRestarts`: Defines the number of restarts allowed per thread.
   - `HeatmapResolution`: Sets the resolution used by the heatmap report.
+  - `LogPerformance`: A boolean parameter to enable/disable the logging of "Performance Metrics" in the default processor log.
+  We recommend enabling it to compare the performance of different devices, models, resolutions, etc.
+  When it's enabled, the processor logs will include the following information every time 100 frames are processed:
+      - Frames per second (FPS)
+      - Average Detector time
+      - Average Classifier time
+      - Average Tracker time
+      - Average Post Processors time
```
Comment (renzodgc) [inline README.md:319]:
```suggestion
      - Average Post Processing time
```

File: libs/cv_engine.py [hunk 2]
```diff
@@ -2,13 +2,20 @@
...
 from libs.classifiers.classifier import Classifier
 from libs.trackers.tracker import Tracker
 from libs.loggers.source_loggers.logger import Logger
 from libs.detectors.detector import Detector
 from libs.source_post_processors.source_post_processor import SourcePostProcessor
 
+
 logger = logging.getLogger(__name__)
+FRAMES_LOG_BATCH_SIZE = 100
+LOG_SECTIONS = ["Detector", "Classifier", "Tracker", "Post Processors"]
```
Comment (renzodgc) [inline libs/cv_engine.py:18]: I'd suggest we change each instance of `Post Processors` to `Post Processing`, I think it makes more sense grammatically

Comment (pgrill) [reply libs/cv_engine.py]: I changed the logging information for the post-processing steps.
Now, each step is log independently.

File: README.md [hunk 3]
```diff
@@ -309,6 +309,17 @@ All the configurations are grouped in *sections* and some of them can vary depen
   - `OccupancyAlertsMinInterval`:  Sets the desired interval (in seconds) between occupancy alerts.
   - `MaxThreadRestarts`: Defines the number of restarts allowed per thread.
   - `HeatmapResolution`: Sets the resolution used by the heatmap report.
+  - `LogPerformance`: A boolean parameter to enable/disable the logging of "Performance Metrics" in the default processor log.
```
Comment (mats-claassen) [inline README.md:312]: What about calling this `LogPerformanceMetrics`?

File: libs/cv_engine.py [hunk 4]
```diff
@@ -50,29 +62,54 @@ def __process(self, cv_image):
...
                     idx = idx + 1
                 else:
                     self.classifier.object_post_process(obj, None, None)
+                classifier_time += (datetime.now() - begin_time).total_seconds()
 
         # Execute post processors
         post_processing_data = {
             "tracks": tracks
         }
+        begin_time = datetime.now()
```
Comment (mats-claassen) [inline libs/cv_engine.py:98]: You can remove this line

File: libs/cv_engine.py [hunk 5]
```diff
@@ -108,3 +146,28 @@ def process_video(self, video_uri):
 
     def stop_process_video(self):
         self.running_video = False
+
+    def reset_log_detail(self):
+        self.log_detail = {}
+        for section in LOG_SECTIONS:
+            if section == "Post processing steps":
```
Comment (mats-claassen) [inline libs/cv_engine.py:153]: Can we put this string in a constant at the top of this file?

**Review by mats-claassen — COMMENTED** [review]
Just thinking whether it would be good to store these metrics in a file. We could name it <device-name>-<model>.txt. I think you will want to compare different models and/or devices and if they are just printed to the console then you will have to copy them manually somewhere to compare. Thoughts?

**Comment by pgrill** [comment]
> Just thinking whether it would be good to store these metrics in a file. We could name it -.txt. I think you will want to compare different models and/or devices and if they are just printed to the console then you will have to copy them manually somewhere to compare. Thoughts?

Yes, I thought about store it in a file. My only concern was that storing it into a file can affect the processor performance for the file access. But I can implement a short version and test it.

File: README.md (hunk 3 above)
Comment (pgrill) [reply README.md]: Agree, I changed

File: libs/cv_engine.py (hunk 4 above)
Comment (pgrill) [reply libs/cv_engine.py]: Removed

File: libs/cv_engine.py (hunk 5 above)
Comment (pgrill) [reply libs/cv_engine.py]: Done

**Comment by pgrill** [comment]
> > Just thinking whether it would be good to store these metrics in a file. We could name it -.txt. I think you will want to compare different models and/or devices and if they are just printed to the console then you will have to copy them manually somewhere to compare. Thoughts?
> 
> Yes, I thought about store it in a file. My only concern was that storing it into a file can affect the processor performance for the file access. But I can implement a short version and test it.

I added now the possibility of storing the performance metrics logs into a CSV file.

**Review by mats-claassen — APPROVED** [review]

**Review by renzodgc — APPROVED** [review]

**Pull request merged**